Port 20w15a pathfinding optimizations

perseus5 opened the pull request:
Mojang decided to remove their usage of streams for brain AI used for Villagers, this ports that change. I am missing the serialization diff, wasn't able to figure that one out. 

For reference: 
20w14a on left, 20w15a on right. 
https://www.diffchecker.com/3R51lUh8

Making this draft while I fix comments/formatting.

perseus5 committed:
Port 20w15a Villager AI optimizations

Spottedleaf commented on Spigot-Server-Patches/0460-Port-20w15a-pathfinding-optimizations.patch:
```diff
@@ -0,0 +1,194 @@
+ import com.mojang.datafixers.util.Pair;
+ import java.util.Collection;
+ import java.util.Iterator;
++import java.util.List;
```
import diff

perseus5 committed:
Async command map building

Reduces lag created on login and world change for sending the command map to client

aikar reviewed (approved):
since its a temporary patch we would drop in 1.16, seems fine to me to not worry about obf helpers and all the normal jazz..

logically seems like the new versions match previous stream version, but def gonna need multiple eyeballs on this to guarantee nothing was missed.

aikar committed:
Allow shutting down server during a watchdog hang gracefully

If the request to shut down the server is received while we are in
a watchdog hang, immediately treat it as a crash and begin the shutdown
process. Shutdown process is now improved to also shutdown cleanly when
not using restart scripts either.

If a server is deadlocked, a server owner can send SIGUP (or any other signal
the JVM understands to shut down as it currently does) and the watchdog
will no longer need to wait until the full timeout, allowing you to trigger
a close process and try to shut the server down gracefully, saving player and
world data.

Previously there was no way to trigger this outside of waiting for a full watchdog
timeout, which may be set to a really long time...

aikar committed:
Restore Optimize Pathfinding patch

aikar committed:
Restore preventing saving bad entities patch to full effect
